airbyte-server/src/main/java/io/airbyte/server/ServerApp.java

```diff
@@ -210,15 +210,50 @@ public static void main(String[] args) throws Exception {
       LOGGER.info(String.format("Setting Database version to %s...", airbyteVersion));
       jobPersistence.setVersion(airbyteVersion);
     }
-    final Optional<String> airbyteDatabaseVersion = jobPersistence.getVersion();
+
+    Optional<String> airbyteDatabaseVersion = jobPersistence.getVersion();
+    if (airbyteDatabaseVersion.isPresent() && !AirbyteVersion
+        .isCompatible(airbyteVersion, airbyteDatabaseVersion.get())
+        && !isDatabaseVersionAheadOfAppVersion(airbyteVersion, airbyteDatabaseVersion.get())) {
+      runAutomaticMigration(configRepository, jobPersistence, airbyteVersion,
+          airbyteDatabaseVersion.get());
+      // After migration, upgrade the DB version
+      airbyteDatabaseVersion = jobPersistence.getVersion();
+    }
 
     if (airbyteDatabaseVersion.isPresent() && AirbyteVersion.isCompatible(airbyteVersion, airbyteDatabaseVersion.get())) {
       LOGGER.info("Starting server...");
       new ServerApp(configRepository, jobPersistence, configs).start();
     } else {
-      LOGGER.info("Start serving version mismatch errors...");
+      LOGGER.info("Start serving version mismatch errors. Automatic migration must have failed");
       new VersionMismatchServer(airbyteVersion, airbyteDatabaseVersion.get(), PORT).start();
     }
   }
 
+  private static void runAutomaticMigration(ConfigRepository configRepository,
+                                            JobPersistence jobPersistence,
+                                            String airbyteVersion,
+                                            String airbyteDatabaseVersion) {
+    LOGGER.info("Running Automatic Migration from version : " + airbyteDatabaseVersion + " to version : " + airbyteVersion);
+    final Path latestSeedsPath = Path.of(System.getProperty("user.dir")).resolve("latest_seeds");
+    LOGGER.info("Last seeds dir: {}", latestSeedsPath);
+    try (RunMigration runMigration = new RunMigration(airbyteDatabaseVersion,
+        jobPersistence, configRepository, airbyteVersion, latestSeedsPath)) {
+      runMigration.run();
+    } catch (Exception e) {
+      LOGGER.error("Automatic Migration failed ", e);
+    }
+  }
+
+  public static boolean isDatabaseVersionAheadOfAppVersion(String airbyteVersion, String airbyteDatabaseVersion) {
```

Comment:
we should have a compareVersion that returns -1, 0, 1 (or an enum) in the existing helper instead of this code.